src/sentry/profiles/task.py

```diff
@@ -225,18 +228,41 @@ def _normalize(profile: Profile, organization: Organization) -> None:
 def _prepare_frames_from_profile(profile: Profile) -> Tuple[List[Any], List[Any]]:
     modules = profile["debug_meta"]["images"]
 
+    # NOTE: the usage of `adjust_instruction_addr` assumes that all
+    # the profilers on all the platforms are walking stacks right from a
+    # suspended threads cpu context
+
     # in the sample format, we have a frames key containing all the frames
     if "version" in profile:
-        stacktraces = [{"registers": {}, "frames": profile["profile"]["frames"]}]
+        frames = profile["profile"]["frames"]
+
+        for stack in profile["profile"]["stacks"]:
+            if len(stack) > 0:
+                # Make a deep copy of the leaf frame with adjust_instruction_addr = False
+                # and append it to the list. This ensures correct behavior
+                # if the leaf frame also shows up in the middle of another stack.
+                first_frame_idx = stack[0]
+                frame = deepcopy(frames[first_frame_idx])
+                frame["adjust_instruction_addr"] = False
+                frames.append(frame)
+                stack[0] = len(frames) - 1
+
+        stacktraces = [{"registers": {}, "frames": [dict(frame) for frame in frames]}]
```

Comment:
Why not transform the frame to a `dict` in the `append` above so you don't have to loop once more?